test/position-managers/PositionManager.modifyLiquidities.t.sol

```diff
@@ -308,4 +336,353 @@ contract PositionManagerModifyLiquiditiesTest is Test, PosmTestSetup, LiquidityF
+        uint256 balance1After = IERC20(Currency.unwrap(currency1)).balanceOf(address(this));
+
+        // The full eth amount was "spent" because some was wrapped into weth and refunded.
+        assertApproxEqAbs(balanceEthBefore - balanceEthAfter, 100 ether, 1 wei);
+        assertApproxEqAbs(balance1Before - balance1After, 100 ether, 1 wei);
+        assertEq(lpm.ownerOf(tokenId), address(this));
+        assertEq(lpm.getPositionLiquidity(tokenId), liquidityAmount);
+        assertEq(_WETH9.balanceOf(address(lpm)), 0);
+        assertEq(address(lpm).balance, 0);
+    }
+
+    function test_wrap_increaseLiquidity_revertsInsufficientBalance() public {
+        // 1 _wrap with more eth than is sent in
+
+        Plan memory planner = Planner.init();
+        // Wrap more eth than what is sent in.
+        planner.add(Actions.WRAP, abi.encode(101 ether));
+
+        bytes memory actions = planner.encode();
+
+        vm.expectRevert(DeltaResolver.InsufficientBalance.selector);
+        lpm.modifyLiquidities{value: 100 ether}(actions, _deadline);
+    }
+
+    function test_unwrap_usingContractBalance() public {
+        // weth-currency1 pool
+        // output: eth, currency1
+        // modifyLiquidities call to mint liquidity weth and currency1
+        // 1 _burn
+        // 2 _take where the weth is sent to the lpm contract
+        // 3 _take where currency1 is sent to the msg sender
+        // 4 _unwrap using contract balance
+        // 5 _sweep where eth is sent to msg sender
+        uint256 tokenId = lpm.nextTokenId();
+
+        uint128 liquidityAmount = LiquidityAmounts.getLiquidityForAmounts(
+            SQRT_PRICE_1_1,
+            TickMath.getSqrtPriceAtTick(wethConfig.tickLower),
+            TickMath.getSqrtPriceAtTick(wethConfig.tickUpper),
+            100 ether,
+            100 ether
+        );
+
+        bytes memory actions = getMintEncoded(wethConfig, liquidityAmount, address(this), ZERO_BYTES);
+        lpm.modifyLiquidities(actions, _deadline);
+
+        assertEq(lpm.getPositionLiquidity(tokenId), liquidityAmount);
+
+        uint256 balanceEthBefore = address(this).balance;
+        uint256 balance1Before = IERC20(Currency.unwrap(currency1)).balanceOf(address(this));
+
+        Plan memory planner = Planner.init();
+        planner.add(
+            Actions.BURN_POSITION, abi.encode(tokenId, MIN_SLIPPAGE_DECREASE, MIN_SLIPPAGE_DECREASE, ZERO_BYTES)
+        );
+        // take the weth to the position manager to be unwrapped
+        planner.add(Actions.TAKE, abi.encode(address(_WETH9), ActionConstants.ADDRESS_THIS, ActionConstants.OPEN_DELTA));
+        planner.add(
+            Actions.TAKE,
+            abi.encode(address(Currency.unwrap(currency1)), ActionConstants.MSG_SENDER, ActionConstants.OPEN_DELTA)
+        );
+        planner.add(Actions.UNWRAP, abi.encode(ActionConstants.CONTRACT_BALANCE));
+        planner.add(Actions.SWEEP, abi.encode(CurrencyLibrary.ADDRESS_ZERO, ActionConstants.MSG_SENDER));
+
```

Comment:
this UNWRAP + SWEEP makes me wonder if its worth giving a recipient param to wrap/unwrap

Reply:
Hm I think we should keep it separate. sometimes we wrap/unwrap and want to send it to the pool manager but that flow should use settle, not just vanilla transfer